### PR TITLE
Port zlib_available_p and zlib_decompress_region

### PR DESCRIPTION
**Pull request by harryfei**
Port decompress related functions.

There is already a test in `test/src/decompress-tests.el`. So I don't add a new one.

### CODE REVIEW

**Comment by shaleh** [comment]
We have been moving the tests to the matching `rust_src` tree when we port.

File: rust_src/src/lib.rs [hunk 1]
```diff
@@ -97,6 +98,7 @@ mod time;
 mod util;
 mod vectors;
 mod windows;
+mod decompress;
```
Comment (shaleh) [inline rust_src/src/lib.rs]: The list is current alphabetically ordered.

File: rust_src/src/decompress.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+        match decoder.read(gap_writer) {
+            // Decompress finished.
+            Ok(0) => {
+                // Delete the compressed data.
+                unsafe { del_range(istart, iend) };
+                return true;
+            }
+
+            // Decompress once successfully.
```
Comment (Wilfred) [inline rust_src/src/decompress.rs]: I don't entirely understand this comment -- could you rephrase it? I think at this point we've decoded the current data fine, but there's still some data left.

File: src/emacs.c [hunk 3]
```diff
@@ -1507,10 +1507,6 @@ Using an Emacs configured with --with-x-toolkit=lucid does not have this problem
       syms_of_lcms2 ();
 #endif
 
-#ifdef HAVE_ZLIB
```
[the remark this answers is not in the document]
Comment (harryfei) [reply src/emacs.c]: I think we can remove HAVE_ZLIB now. I will remove them in this PR update.

**Comment by harryfei** [comment]
Fixed those:
* Move decompress tests to rust_src/src
* Remove all HAVE_ZLIB
* Change some comments
* Fixed the failed test
* Fixed mod declare order in `lib.rs`

@Wilfred @shaleh

File: rust_src/src/decompress.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+fn create_buffer_decoder<'a>(buffer: &'a [u8]) -> Box<Read + 'a> {
+    let magic_number = buffer[0];
+
+    match magic_number {
+        // Zlib
+        0x78 => Box::new(ZlibDecoder::new(buffer)),
+        // Gzlib
+        0x1F => Box::new(GzDecoder::new(buffer)),
```
Comment (shaleh) [inline rust_src/src/decompress.rs]: I think the comment reads better this way.

`// Assume the data is raw if neither zlib nor gzip header can be found.`

File: rust_src/src/decompress.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    let old_pt = current_buffer.pt();
+    current_buffer.set_pt_both(charpos, bytepos);
+
+    let compressed_buffer = unsafe {
+        slice::from_raw_parts(
+            current_buffer.byte_pos_addr(istart),
+            (iend - istart) as usize,
+        )
+    };
+
```
Comment (shaleh) [inline rust_src/src/decompress.rs]: Spelling. The word is 'decompressor'.

File: rust_src/src/decompress.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        match decoder.read(gap_writer) {
+            // Decompress all data finished.
+            Ok(0) => {
+                // Delete the compressed data.
+                unsafe { del_range(istart, iend) };
+                return true;
+            }
+
+            // Decompress one batch data successfully.
```
Comment (shaleh) [inline rust_src/src/decompress.rs]: Typo. The word should be 'remaining'.

**Comment by shaleh** [comment]
Looks good. I just left some grammar and spelling comments.

File: rust_src/src/decompress.rs (hunk 6 above)
Comment (harryfei) [reply rust_src/src/decompress.rs]: Fixed.

File: rust_src/src/decompress.rs (hunk 5 above)
Comment (harryfei) [reply rust_src/src/decompress.rs]: Fixed it.

File: rust_src/src/decompress.rs (hunk 4 above)
Comment (harryfei) [reply rust_src/src/decompress.rs]: Fixed it. Thanks.
